fgspectra/model.py

```diff
@@ -0,0 +1,127 @@
+from abc import ABC, abstractmethod
+import types
+import inspect
+import yaml
+import numpy as np
+
+class Model(ABC):
+    """ Abstract class for model definition
+
+    A model is a class that has one purpose: evaluating the model. What
+    evaluation means is defined by the `eval` method that any child class has to
+    override.
+
+    If the `eval` method of a hypotetical ``Child`` class calls the `eval`
+    method of other `Model`s, it is likely that ``Child`` has also to override
+    `set_defaults`, `defaults` and `_get_repr`
+    """
+
+    def __init__(self, **kwargs):
+        """ You can set defaults at construction time """
+        self.set_defaults(**kwargs)
+
+    def set_defaults(self, **kwargs):
+        """ Change the defaults of the evaluation
+
+        Call ``set_defaults`` with the same arguments (or a subset) of ``eval``
+        (or ``__call__``).
+        This will change the default value of those arguments in ``eval``. If
+        `kwargs` has keys that are not arguments of `eval`, these keys are
+        ignored.
+
+        Examples
+        --------
+
+        >>> class PrintABC(Model):
+        ...     def eval(self, a=1, b=2, c=3):
+        ...         print(a, b, c)
+        ...
+        >>> print_abc = PrintABC()
+        >>> print_abc.eval()
+        1 2 3
+        >>> print_ab10c = PrintABC()
+        >>> print_ab10c.set_defaults(b=10)
+        >>> print_ab10c.eval()  # The default of b ahs changed
+        1 10 3
+        >>> print_abc.eval()  # Without affecting other instances
+        1 2 3
+
+        """
+        # This is a dirty business, and has limitations. The most important
+        # known one is that it breaks with decorated eval
+        # To change the default values of the eval method
+        # we have to bind a new method to the instance
+        # (not the class, otherwise we change the defaults of all the instances)
+
+        if not kwargs:  # No default to be replaced
+            return
+
+        # Make a deep copy of the eval function, assuming no variadic arguments
+        func = self.eval.__func__
```

Comment:
If eval is actually wrapped, this does not pick up the right function (i.e. the inner most). If we pick the innermost, then we have to apply the decorator to the new instance. I didn't handle this at the moment, that is why decorators brake this method

Reply:
I'm glad you've put a bunch of comments here to prevent users from adding a decorator (vast majority will not, probably). Surely there exists a way to do this better? It might be a good stackoverflow question?